app/screens/menu_action_selector/menu_action_selector.js

```diff
@@ -0,0 +1,268 @@
+        data: PropTypes.arrayOf(PropTypes.object),
+        dataSource: PropTypes.string,
+        onSelect: PropTypes.func.isRequired,
+        currentTeamId: PropTypes.string.isRequired,
+        loadMoreRequestStatus: PropTypes.string,
+        searchRequestStatus: PropTypes.string,
+        actions: PropTypes.shape({
+            getProfiles: PropTypes.func.isRequired,
+            getChannels: PropTypes.func.isRequired,
+            searchProfiles: PropTypes.func.isRequired,
+            searchChannels: PropTypes.func.isRequired,
+        }),
+    };
+
+    constructor(props) {
+        super(props);
+
+        this.searchTimeoutId = 0;
+
+        let data = [];
+        if (!props.dataSource) {
+            data = props.data;
+        }
+
+        const needsLoading = props.dataSource === 'users' || props.dataSource === 'channels';
+
+        this.state = {
+            next: needsLoading,
+            page: 0,
+            data,
+            searching: false,
+            showNoResults: false,
+            term: '',
+            isLoading: needsLoading,
+            prevLoadMoreRequestStatus: RequestStatus.STARTED,
+        };
+
+        props.navigator.setOnNavigatorEvent(this.onNavigatorEvent);
+    }
+
+    componentDidMount() {
+        InteractionManager.runAfterInteractions(() => {
+            if (this.props.dataSource === 'users') {
+                this.props.actions.getProfiles().then(() => this.setState({isLoading: false}));
+            } else if (this.props.dataSource === 'channels') {
+                this.props.actions.getChannels(this.props.currentTeamId).then(() => this.setState({isLoading: false}));
+            }
+        });
+    }
+
+    componentDidUpdate(prevProps) {
+        if (this.props.theme !== prevProps.theme) {
+            setNavigatorStyles(this.props.navigator, this.props.theme);
+        }
+    }
+
+    cancelSearch = () => {
+        this.setState({
+            searching: false,
+            isLoading: false,
+            term: '',
+            page: 0,
+            data: filterPageData(this.props.data, 0),
+        });
+    };
+
+    close = () => {
+        this.props.navigator.pop({animated: true});
+    };
+
+    handleRowSelect = (id, selected) => {
+        this.props.onSelect(selected);
+        this.close();
+    };
+
+    loadMore = async () => {
+        const {actions, loadMoreRequestStatus, currentTeamId, dataSource} = this.props;
+        const {next, searching} = this.state;
+        let {page} = this.state;
+
+        if (loadMoreRequestStatus !== RequestStatus.STARTED && next && !searching) {
+            page = page + 1;
+
+            let results;
+            if (dataSource === 'users') {
+                results = await actions.getProfiles(page, General.PROFILE_CHUNK_SIZE);
+            } else if (dataSource === 'channels') {
+                results = await actions.getChannels(currentTeamId, page, General.PROFILE_CHUNK_SIZE);
+            } else {
+                return;
```

Comment:
There's something about returning only in an else case that makes me uncomfortable. Maybe just add an early return before this if `dataSource` isn't one of those two values?

Reply:
Eeehhh I agree it's not great to look at but it does exactly what I need it to and I'd rather not duplicate the checks I'm already doing

Reply:
I wouldn't be worried about the extra check since it's just a simple string comparison and the compiler would probably deal with it anyway. I'm fine leaving it though.